source/load-balancers/load-balancers.rst

```diff
@@ -0,0 +1,553 @@
+service behind the load balancer. The rules for pinning to a connection and releasing
+a pinned connection are the same as those for server pinning in non-load balanced sharded
+transactions as described in `When to unpin <../transactions/transactions.rst#when-to-unpin>`__.
+
+Connection Tracking
+^^^^^^^^^^^^^^^^^^^
+
+The driver connection pool MUST track the purpose for which connections are checked out
+in the following 3 categories:
+
+- Connections checked out for cursors
+- Connections checked out for transactions
+- Connections checked out for operations not falling under the previous 2 categories
+
+When the connection pool's :code:`maxPoolSize` is reached and the pool times out waiting
+for a new connection the :code:`WaitQueueTimeoutError` MUST include a new detailed message,
+"Timeout waiting for connection from the connection pool. maxPoolSize: n, connections
+in use by cursors: n, connections in use by transactions: n, connections in use by other
+operations: n". A :code:`ConnectionCheckOutFailedEvent` MUST also be published with the
+same message.
+
+
+Error Handling
+--------------
+
+Initial Handshake Errors
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+If there is a network error or timeout on the connection before the hello or legacy hello completes,
+the driver MUST NOT make any changes to the :code:`TopologyDescription` or mark the server as Unknown.
+The driver MUST also close the associated connection. If the network error occurred on the handshake's
+:code:`hello` command, the connection pool MUST NOT be cleared. If the network error occurred after
+the handshake's hello command, the connection POOL must clear close all connections with the matching
+:code:`serviceId`.
+
+Post-Handshake Errors
+^^^^^^^^^^^^^^^^^^^^^^
+
+When the driver is in load balanced mode and encounters any error described here as a state
+change error, the driver MUST NOT make any changes to the :code:`TopologyDescription` or the
+:code:`ServerDescription` of the load balancer (i.e. it MUST NOT mark the load balancer as Unknown).
+If the error requires the connection pool to be cleared, the driver MUST only clear connections
+with the same :code:`serviceId` as the connection which errored.
+
+
+Events
+------
+
+When in load balancer mode the driver MUST now include the :code:`serviceId` in the
+:code:`CommandStartedEvent`, :code:`CommandSucceededEvent`, and
+:code:`CommandFailedEvent`. The driver MAY decide how to expose this information.
+Drivers that have a :code:`ConnectionId` object for example, MAY choose to provide a
+:code:`serviceId` in that object. The :code:`serviceId` field is only present when
+in load balancer mode and connected to a service that is behind a load balancer.
+
+Additionally the :code:`PoolClearedEvent` MUST also contain a :code:`serviceId`
+field.
+
+
+Implementation Notes
+--------------------
+
+All connections MUST track a generation number. For non-load balanced deployments, the connection's
```

Comment:
Most of the information in this section is duplicated across parts of this spec and other specs like CMAP. Can we disperse it into the relevant places and basically remove the "Implementation Notes" section completely?

Reply:
I've removed the implementation notes now, and stuck with the duplicated information in the CMAP spec and added the error handling notes in that spec below it.